### ISSUE TITLE
Compiled ObjectManger doesn't properly create objects of class that are outside of registered components paths

### ISSUE THREAD

**Issue by mmenozzi** [magento/magento2#9760]
### Preconditions

1. Magento 2.1.* (but I think even for 2.0.*)
2. PHP 5.6.* and 7.0.*

### Steps to reproduce
1. Install Magento
2. `composer require webgriffe/lib-quipago` (this is only an example, as you'll see the problem could be with any library which is not a Magento component)
3. In a controller action class of your choice add the class `Webgriffe\LibQuiPago\PaymentInit\UrlGenerator` as a dependency (so add it as a constructor argument)
4. Navigate to the action via browser and notice that it should work as expected
5. Run the `php bin/magento setup:di:compile` command
6. Navigate to the action via browser again

### Expected result
1. The action should work again as expected because we only added a dependency to the action class which does nothing at all 

### Actual result
1. You should get the following error:

```
Fatal error: Uncaught TypeError: Argument 1 passed to Webgriffe\LibQuiPago\PaymentInit\UrlGenerator::__construct() must implement interface Psr\Log\LoggerInterface, instance of Magento\Framework\ObjectManager\ObjectManager given
```

### Additional details
This happens because one of the operations of the `php bin/magento setup:di:compile` command is to fetch all the constructor arguments of available classes and serialize them in the `var/di/*.ser` files. The `setup:di:compile` command searches for available classes **only in the paths of the registered Magento components**. An external library, installed with Composer, is not a Magento registered component, so its classes are not processed.
Then, when the `var/di/*.ser` files are available, the ObjectManager uses the `Compiled` definitions. In this mode when the ObjectManager is asked to create an instance of a certain class it looks for its constructor arguments in the `var/di/*.ser` files. But, as explained above, classes of external libraries are not processed so the ObjectManager doesn't know what dependencies to inject. In such situation the ObjectManager tries to instantiate that class by passing the ObjectManager itself. Then, if the dependency class has a type hinting on the first constructor argument, you'll get the error above.
Indeed, if you delete all files in the `var/di` directory, the controller action will work again.

**Comment by yellow13** [comment]
Facing the same problem with a custom CLI command (which also cause the CLI stop working until /var/di* is deleted).

Spending hours searching on the internet and trying solutions thinking that was something wrong with my code and finally found this bug report.

I hope this will be fixed soon and that a workaround will be available since I, and who is affected by this bug, can't run setup:di:compile.

**Comment by liamfirth** [comment]
Still not fixed in 2.1.7. Worked around by calling class directly `new \GuzzleHttp\Client();`

**Comment by dmanners** [comment]
Hi @mmenozzi I think this is not actually a bug but a lack di.xml in your module. From what I see you have an entry point in a Magento controller for your class `Webgriffe\LibQuiPago\PaymentInit\UrlGenerator`. So Magento needs to have the right di.xml to build this controller and all the classes that are injected into the controller. In this case we have your class `Webgriffe\LibQuiPago\PaymentInit\UrlGenerator` from your library. Which in itself needs `Psr\Log\LoggerInterface`. So when you run di:compile Magento's di tool will try to find all the implementations it needs for the interfaces defined. In this case I think in your Magento module's di.xml you should be able to add `<preference for="Psr\Log\LoggerInterface" type="Your\Implementation\Of\Choice" />`.

That being said I am fairly sure that there should be a defined preference for `Psr\Log\LoggerInterface` in most if not all 2.1.* versions. I just tried this on Magento CLI version 2.1.7 by requiring your module (version ^1.0) and then making a class that required your class `Webgriffe\LibQuiPago\PaymentInit\UrlGenerator` it all compiled correctly for me. Did you get any further with this issue yourself?

**Comment by mmenozzi** [comment]
As a workaround for this issue you can create an empty "adapter" class which simply extends the one that lives in the non-magento-component library. The adapter class must be inside the Magento module which uses the library.

**Comment by magento-engcom-team** [comment]
@mmenozzi, thank you for your report.
The issue is already fixed in develop branch, 2.2.0

**Comment by mmenozzi** [comment]
Hi @magento-engcom-team you're welcome.
I can confirm that the issue is now solved in the version 2.2.0.

**Comment by magento-engcom-team** [comment]
@mmenozzi Thank you for the update.
